### PR TITLE
Follow load statements for symbols and functions

### PR DESCRIPTION
**Pull request by nicksieger**
- analysis: extract NodesToContent and availableSymbols
- query: decouple from document package; extract DocumentContent
- document: move parsing to manager Read/Write
- document: associate functions and symbols to a document
- document: pre-parse symbols at document creation time
- document: build out manager internal parse routines
- document: add diagnostics and publish them on DidChange
- document: attempt to canonicalize URIs
- document: stub out processLoads and circular loading
- query: LoadStatements
- document: processLoads parses and follows load statements
- document: add uri
- document: resolve relative paths to doc.uri or workspace folder
- server: ensure empty slice to clear diagnostics

### CODE REVIEW

File: pkg/document/document.go [hunk 1]
```diff
@@ -1,60 +1,298 @@
...
+	}
+	for fn := range d.functions {
+		doc.functions[fn] = d.functions[fn]
+	}
+	return doc
+}
+
+func (d *document) processLoads(ctx context.Context, m *Manager) {
+	for i, load := range d.loads {
+		if load.File == "" {
```
Comment (landism) [inline pkg/document/document.go:145]: we should probably also skip files that start with "ext://" (or maybe contain "://" at all)

(fwiw, if referring to "ext://" as a "file" feels weird - the starlark spec resolves this by referring to the first arg as the "module")

File: pkg/document/manager.go [hunk 2]
```diff
@@ -36,31 +48,112 @@ func WithNewDocumentFunc(newDocFunc NewDocumentFunc) ManagerOpt {
...
+	defer func() {
+		if err == nil {
+			// Always return a copy of the document
+			doc = doc.Copy()
+		}
+		m.mu.Unlock()
+	}()
+	u = canonicalURI(u, m.root)
+
+	// TODO(siegs): check staleness for files read from disk?
```
Comment (nicksieger) [inline pkg/document/manager.go:122]: Wondering if this should be handled now or in a future PR -- as written this caches documents and they won't be updated unless separately edited in the same editor that is talking to the current LSP server. If you make edits outside of the LSP client editor, they won't be reflected.

File: pkg/document/document.go [hunk 3]
```diff
@@ -1,60 +1,298 @@
...
+	sitter "github.com/smacker/go-tree-sitter"
+	"go.lsp.dev/protocol"
+	"go.lsp.dev/uri"
+
+	"github.com/tilt-dev/starlark-lsp/pkg/query"
+)
+
+type LoadStatement struct {
+	File        string
+	Vars        [][2]string
```
Comment (landism) [inline pkg/document/document.go:19]: "Symbols" might be better here - it's easy to read "Vars" as excluding functions

File: pkg/document/document.go [hunk 4]
```diff
@@ -1,60 +1,298 @@
...
+				}
+			default:
+				diagnostics = append(diagnostics, notAString(va))
+			}
+		}
+	} else {
+		diagnostics = append(diagnostics, protocol.Diagnostic{
+			Range:    query.NodeRange(n),
+			Severity: protocol.DiagnosticSeverityWarning,
+			Message:  "load statement did not specify any variables to import",
```
Comment (landism) [inline pkg/document/document.go:266]:
```suggestion
			Message:  "load statement did not specify any symbols to import",
```

File: pkg/document/document.go [hunk 5]
```diff
@@ -1,60 +1,298 @@
...
+		})
+	}
+	return load, diagnostics
+}
+
+func unquote(s string) string {
+	s, _ = strconv.Unquote(`"` + strings.Trim(s, s[0:1]) + `"`)
```
Comment (landism) [inline pkg/document/document.go:273]: Is there a way to get the actual, unquoted string from treesitter? Some of these starlark string literals seem like they'd fail here:
1. `r"foo"`
2. `"""foo"""`
3. `r"""foo"""` (apparently starlark-go accepts this, though I read the spec as not allowing it)
4. `'fo"o'`
5. `'fo\'o'`

(also maybe a good reason to handle the error)

**Review by landism — COMMENTED** [review]
I'm excited for this! I'm still going through the PR but figured I might as well publish comments before stopping for the day.

File: pkg/document/document.go (hunk 3 above)
Comment (nicksieger) [reply pkg/document/document.go:19]: Yeah, I like that better

File: pkg/document/document.go (hunk 5 above)
Comment (nicksieger) [reply pkg/document/document.go:273]: Ooh, yeah, those are good examples. I didn't find a way to get it from treesitter initially, but I'll have another look around and see what I can find out. At a minimum I can add tests for these, I admittedly didn't spend enough time thinking through edge cases for this.

Comment (landism) [reply pkg/document/document.go:273]: it occurred to me that getting docstrings is the same problem, so we could just reuse that solution, but it looks like it has similar problems:
https://github.com/tilt-dev/starlark-lsp/blob/c42183ababe2754eea0cf93c5050c527ce547235/pkg/query/symbol.go#L40

File: pkg/query/conv.go [hunk 6]
```diff
@@ -1,10 +1,56 @@
...
+	for {
+		switch n.Type() {
+		case NodeTypeModule,
+			NodeTypeBlock,
+			NodeTypeExpressionStatement:
+			n = n.Child(0)
+		case NodeTypeString:
+			break done
+		default:
+			return ""
```
Comment (nicksieger) [inline pkg/query/conv.go:21]: Should this return an error instead? If you're not passing a thing containing a string literal I feel like this is the best we can do.

File: pkg/query/conv.go [hunk 7]
```diff
@@ -1,10 +1,56 @@
...
+		}
+	}
+
+	startDelim := n.Child(0)
+	endDelim := n.Child(int(n.ChildCount() - 1))
+	byteoffset := startDelim.EndByte()
+	bytes := []byte{}
+
+	for i := 1; i < int(n.ChildCount()-1); i++ {
+		escape := n.Child(i)
```
Comment (landism) [inline pkg/query/conv.go:31]: Can you add a comment explaining `n`'s children? I'm struggling to read this even with consulting the treesitter grammar. Or maybe this is illustratable with a treesitter playground link?

Presumably the first and last children are the quotes themselves. This code is referring to every child in between as an `escape`, which leaves me confused where the `foo` in `"foo"` would be handled, since that doesn't have an escape. Do we have an iteration where escape == "foo" (or maybe one iteration per char)?

or I guess unescaped chars are handled by the next if statement because they don't belong to a node at all?

File: pkg/query/conv.go [hunk 8]
```diff
@@ -1,10 +1,56 @@
...
+		if byteoffset < escape.StartByte() {
+			bytes = append(bytes, input[byteoffset:escape.StartByte()]...)
+			byteoffset = escape.EndByte()
+		}
+		escseq := string(input[escape.StartByte():escape.EndByte()])
```
Comment (landism) [inline pkg/query/conv.go:36]: fwiw, I think this is just `escape.Content(input)`

**Review by landism — APPROVED** [review]
whew, nice!

File: pkg/document/manager.go [hunk 9]
```diff
@@ -73,6 +166,44 @@ func (m *Manager) Keys() []uri.URI {
 	return keys
 }
 
+func (m *Manager) parseSetup() {
+	m.parseState = make(map[uri.URI]Document)
+}
```
[the remark this answers is not in the document]
Comment (nicksieger) [reply pkg/document/manager.go:174]: I'll try that on, that makes sense to me.

File: pkg/analysis/completion.go [hunk 10]
```diff
@@ -142,6 +126,43 @@ func (a *Analyzer) completeExpression(doc document.Document, nodes []*sitter.Nod
 	return symbols
 }
 
+// Returns a list of available symbols for completion as follows:
+// - If in a function argument list, include keyword args for that function
+// - Add symbols in scope for the node at point
+// - Add document symbols, excluding symbols after the node if the node is in the module scope
```
[the remark this answers is not in the document]
Comment (nicksieger) [reply pkg/analysis/completion.go:133]: Does this 9173887bd5add7c1b886b6e917b33d8b5aefc65f help?

File: pkg/document/document.go [hunk 11]
```diff
@@ -1,60 +1,323 @@
...
 
-func NewDocument(input []byte, tree *sitter.Tree) Document {
-	return document{
+func NewDocument(u uri.URI, input []byte, tree *sitter.Tree) Document {
+	return &document{
+		uri:   u,
 		input: input,
 		tree:  tree,
 	}
 }
```
[the remark this answers is not in the document]
Comment (nicksieger) [reply pkg/document/document.go:48]: That sounds good to me! I think the only thing was extra computations to find symbols, but it's probably negligible in tests. ea1c80f038aad77b0337a756dae2cd875914c72e

File: pkg/query/conv.go [hunk 12]
```diff
@@ -1,10 +1,56 @@
...
+		if escseq == "\\\n" {
+			// ignore backslash at the end of a line per Starlark spec
+			escseq = "\n"
```
[the remark this answers is not in the document]
Comment (nicksieger) [reply pkg/query/conv.go]: Ah, yes, good catch.
0158d90ad21ed959bd5f9b8fbb46564e6d964254

File: pkg/query/conv.go (hunk 7 above)
Comment (nicksieger) [reply pkg/query/conv.go:31]: Yeah, there are no nodes representing the unescaped characters in the string. Let me see if I can explain that for posterity.

File: pkg/query/conv.go (hunk 8 above)
Comment (nicksieger) [reply pkg/query/conv.go:36]: Indeed

File: pkg/query/conv.go (hunk 7 above)
Comment (nicksieger) [reply pkg/query/conv.go:31]: 9cde4ca79b810596815d90eb582a52e08b9ffffe

Comment (nicksieger) [reply pkg/query/conv.go:56]: 9cde4ca79b810596815d90eb582a52e08b9ffffe

File: pkg/analysis/completion.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (nicksieger) [reply pkg/analysis/completion.go]: Ah, no. Added more explanation in 03e1e133d5302ffb505245a6f267164fe7e8d212. Not sure how much explanation is necessary in comments, but what I mean is this. Consider:
```python
a = 1
# <- point 1
def b():
  # <- point 2
  return 2
b()
# <- point 3
c = 3
```
1. At point 1, the visible symbols should only be `a`.
2. At point 2, the visible symbols should be `a`, `b`, and `c` since we're inside a function.
3. At point 3, the visible symbols should be `a` and `b`, but not `c`.

Not sure what the Starlark spec says about scoping but this is the behavior I have noticed. Let me know if you think that's not quite right.

[the remark this answers is not in the document]
Comment (landism) [reply pkg/analysis/completion.go]: yeah, if you enter this at point 3, it prints 3:
```
f = lambda: print(c)
f()
```

Maybe a lambda is actually the only exception to that and we could special case it? I'd think I'd rather simplify and cut the special case entirely rather than add a special case within a special case.

fwiw, the python extension just completes it:
image

The starlark spec says it's a static error to refer to a variable name that's not bound at all, and a dynamic error to refer to a variable name before it's bound
